### ISSUE TITLE
Unable to reuse package if patch is changed/removed

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#30557]
### Steps to reproduce

#29239 changed the checksum for several packages in Spack. Now, when I try to reuse existing packages in my environment, the environment fails to update.

To reproduce this: 
1. install any package with a patch
2. remove that patch from Spack
3. try to reuse that package in an environment.

This may be the same issue I reported in #19649. Also see #27902. These issues were supposedly fixed by @tgamblin in #19672 and @becker33 in #29574. This is quite problematic because it prevents users from upgrading to newer versions of Spack where patch may have changed or been removed. I think the right approach would be to use the historic `package.py` and patch file installed with the package instead of the one currently in Spack. I believe I talked with @alalazo about this years ago.

### Error message

```console
$ spack -d -e . install
==> [2022-05-08-21:59:22.685893] Reading config file /Users/Adam/spack/etc/spack/defaults/config.yaml
==> [2022-05-08-21:59:22.864177] Reading config file /Users/Adam/.spack/spack.yaml
==> [2022-05-08-21:59:22.891332] Using environment '/Users/Adam/.spack'
==> [2022-05-08-21:59:22.897748] Imported install from built-in commands
==> [2022-05-08-21:59:22.902179] Imported install from built-in commands
==> [2022-05-08-21:59:23.276685] Installing environment /Users/Adam/.spack
==> [2022-05-08-21:59:23.276848] Assessing installation status of environment packages
==> [2022-05-08-21:59:23.277650] Reading config file /Users/Adam/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-05-08-21:59:23.295196] DATABASE LOCK TIMEOUT: 3s
==> [2022-05-08-21:59:23.295313] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-05-08-21:59:24.227303] All of the packages are already installed
==> [2022-05-08-21:59:25.844150] 0 specs need to be overwritten
==> [2022-05-08-21:59:25.844269] Initializing the build queue from the build requests
==> [2022-05-08-21:59:25.844333] Ensure all dependencies know all dependents across specs
==> [2022-05-08-21:59:26.015630] Reading config file /Users/Adam/spack/etc/spack/defaults/repos.yaml
==> [2022-05-08-21:59:29.606909] NoSuchPatchError: Couldn't find patch for package builtin.py-torch with sha256: e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9
==> [2022-05-08-21:59:29.607427] Error: Couldn't find patch for package builtin.py-torch with sha256: e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9
Traceback (most recent call last):
  File "/Users/Adam/spack/lib/spack/llnl/util/lang.py", line 193, in _memoized_function
    return func.cache[key]
KeyError: (py-torch@1.11.0%apple-clang@12.0.0+breakpad+caffe2~cuda~distributed+fbgemm+kineto+metal+mkldnn+nnpack+numpy+onnx_ml+openmp+qnnpack~rocm~test+xnnpack patches=e37afff arch=darwin-catalina-ivybridge ^apple-libuuid@1353.100.2%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^bzip2@1.0.8%apple-clang@12.0.0~debug~pic+shared arch=darwin-catalina-ivybridge ^eigen@3.4.0%apple-clang@12.0.0~ipo build_type=RelWithDebInfo arch=darwin-catalina-ivybridge ^expat@2.4.6%apple-clang@12.0.0~libbsd arch=darwin-catalina-ivybridge ^fxdiv@2020-04-17%apple-clang@12.0.0~ipo build_type=RelWithDebInfo arch=darwin-catalina-ivybridge ^gdbm@1.19%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^gettext@0.21%apple-clang@12.0.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=darwin-catalina-ivybridge ^intel-mkl@2020.0.166%apple-clang@12.0.0~ilp64+shared threads=none arch=darwin-catalina-ivybridge ^libffi@3.4.2%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^libiconv@1.16%apple-clang@12.0.0 libs=shared,static arch=darwin-catalina-ivybridge ^libxml2@2.9.12%apple-clang@12.0.0~python arch=darwin-catalina-ivybridge ^libyaml@0.2.5%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^llvm-openmp@12.0.1%apple-clang@12.0.0~ipo~multicompat build_type=RelWithDebInfo arch=darwin-catalina-ivybridge ^ncurses@6.2%apple-clang@12.0.0~symlinks+termlib abi=none arch=darwin-catalina-ivybridge ^openssl@1.1.1m%apple-clang@12.0.0~docs certs=system arch=darwin-catalina-ivybridge ^protobuf@3.18.0%apple-clang@12.0.0+shared build_type=Release arch=darwin-catalina-ivybridge ^psimd@2020-05-17%apple-clang@12.0.0~ipo build_type=RelWithDebInfo arch=darwin-catalina-ivybridge ^pthreadpool@2021-04-13%apple-clang@12.0.0~ipo build_type=RelWithDebInfo arch=darwin-catalina-ivybridge ^py-future@0.18.2%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^py-numpy@1.22.2%apple-clang@12.0.0+blas+lapack patches=873745d arch=darwin-catalina-ivybridge ^py-protobuf@3.17.3%apple-clang@12.0.0~cpp arch=darwin-catalina-ivybridge ^py-pybind11@2.6.2%apple-clang@12.0.0~ipo build_type=RelWithDebInfo arch=darwin-catalina-ivybridge ^py-pyyaml@6.0%apple-clang@12.0.0+libyaml arch=darwin-catalina-ivybridge ^py-setuptools@59.4.0%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^py-six@1.16.0%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^py-tqdm@4.56.2%apple-clang@12.0.0~notebook~telegram arch=darwin-catalina-ivybridge ^py-typing-extensions@3.10.0.2%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^python@3.9.10%apple-clang@12.0.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=darwin-catalina-ivybridge ^readline@8.1%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^sqlite@3.37.2%apple-clang@12.0.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=darwin-catalina-ivybridge ^tar@1.34%apple-clang@12.0.0 arch=darwin-catalina-ivybridge ^xz@5.2.5%apple-clang@12.0.0~pic libs=shared,static arch=darwin-catalina-ivybridge ^zlib@1.2.11%apple-clang@12.0.0+optimize+pic+shared arch=darwin-catalina-ivybridge,)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 910, in main
    return _main(argv)
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 865, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 893, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 550, in _invoke_command
    return_val = command(parser, args)
  File "/Users/Adam/spack/lib/spack/spack/cmd/install.py", line 356, in install
    env.install_all(**kwargs)
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 1558, in install_all
    self.install_specs(None, **install_args)
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 1613, in install_specs
    self.regenerate_views()
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 1374, in regenerate_views
    view.regenerate(concretized_root_specs)
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 539, in regenerate
    new_root = self._next_root(specs)
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 434, in _next_root
    content_hash = self.content_hash(specs)
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 442, in content_hash
    ('specs', [(spec.full_hash(), spec.prefix) for spec in sorted(specs)])
  File "/Users/Adam/spack/lib/spack/spack/environment/environment.py", line 442, in <listcomp>
    ('specs', [(spec.full_hash(), spec.prefix) for spec in sorted(specs)])
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1835, in full_hash
    return self._cached_hash(ht.full_hash, length)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1793, in _cached_hash
    hash_string = self._spec_hash(hash)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1772, in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1950, in to_node_dict
    hash_tuple = (hash.name, dspec.spec._cached_hash(hash))
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1793, in _cached_hash
    hash_string = self._spec_hash(hash)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1772, in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1935, in to_node_dict
    package_hash = self.package_hash()
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1801, in package_hash
    return self._cached_hash(ht.package_hash)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1793, in _cached_hash
    hash_string = self._spec_hash(hash)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 1771, in _spec_hash
    return hash.override(self)
  File "/Users/Adam/spack/lib/spack/spack/hash_types.py", line 62, in <lambda>
    override=lambda s: s.package.content_hash())
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 1703, in content_hash
    for p in self.spec.patches)
  File "/Users/Adam/spack/lib/spack/llnl/util/lang.py", line 195, in _memoized_function
    ret = func(*args, **kwargs)
  File "/Users/Adam/spack/lib/spack/spack/spec.py", line 3710, in patches
    patch = index.patch_for_package(sha256, self.package)
  File "/Users/Adam/spack/lib/spack/spack/patch.py", line 367, in patch_for_package
    raise NoSuchPatchError(
spack.patch.NoSuchPatchError: Couldn't find patch for package builtin.py-torch with sha256: e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9
```

### Information on your system

* **Spack:** 0.18.0.dev0 (0dd9e5c86fa1c67e72bc64966ccdaceb4b44c193)
* **Python:** 3.9.10
* **Platform:** darwin-catalina-ivybridge
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
Note: this likely also affects packages whose patches are removed from the repo.

**Comment by adamjstewart** [comment]
Found a leftover TODO that alludes to this issue: https://github.com/spack/spack/blob/v0.17.2/lib/spack/spack/spec.py#L3419,L3420

**Comment by adamjstewart** [comment]
For anyone else who encounters this bug, I found a workaround. If you create a fake `patch_dict` and only update it if the patch is actually found, you can get your installation to succeed:
```patch
diff --git a/lib/spack/spack/patch.py b/lib/spack/spack/patch.py
index 33f9fa57eb..7c96d05838 100644
--- a/lib/spack/spack/patch.py
+++ b/lib/spack/spack/patch.py
@@ -362,21 +362,22 @@ def patch_for_package(self, sha256, pkg):
         we have information about the package's location in its repo.
 
         """
-        sha_index = self.index.get(sha256)
-        if not sha_index:
-            raise NoSuchPatchError(
-                "Couldn't find patch for package %s with sha256: %s"
-                % (pkg.fullname, sha256))
+        patch_dict = {
+            'owner': pkg.fullname,
+            'level': 1,
+            'working_dir':
+            '.',
+            'url': 'https://ftpmirror.gnu.org/bash/bash-5.1-patches/bash51-001',
+        }
 
-        # Find patches for this class or any class it inherits from
-        for fullname in pkg.fullnames:
-            patch_dict = sha_index.get(fullname)
-            if patch_dict:
-                break
-        else:
-            raise NoSuchPatchError(
-                "Couldn't find patch for package %s with sha256: %s"
-                % (pkg.fullname, sha256))
+        sha_index = self.index.get(sha256)
+        if sha_index:
+            # Find patches for this class or any class it inherits from
+            for fullname in pkg.fullnames:
+                new_dict = sha_index.get(fullname)
+                if new_dict:
+                    patch_dict = new_dict
+                    break
 
         # add the sha256 back (we take it out on write to save space,
         # because it's the index key)
```
Hopefully we can find a better solution than this hack.

**Comment by haampie** [comment]
@tgamblin spack uninstall is completely broken for me.

```
$ spack uninstall --all %gcc@4.5.0
==> The following packages will be uninstalled:

    -- test-debian6-core2 / gcc@4.5.0 -------------------------------
    snduugz libelf@0.8.13

==> Do you want to proceed? [y/N] y
==> Warning: Missing a source id for sirius@develop
==> Warning: Missing a source id for q-e-sirius@develop-ristretto
==> Error: Couldn't find patch for package builtin.binutils with sha256: 87c08dba4b77618f96e9a0e7f34f1f69f548c7f8299446898b6e29c724947b69
```

I fails/warns when:

1. The database has a package with a missing patch
2. The database has a package that no longer exists
3. The database has a package with a version no longer registered

**Comment by Paul-Ferrell** [comment]
This blocking my progress on the next TCE release, so it's a pretty high priority bug from my perspective.

**Comment by adamjstewart** [comment]
@Paul-Ferrell can you see if this issue was fixed by #30678? That fixed it for me.

**Comment by markyoder** [comment]
Yup; I'm getting this too. Moving (rebuilding) a working environment from development to my production area (different `spack` installations and:
```
(base) [myoder96@sh03-04n57 /home/groups/s-ees/share/cees/spack_cees] (job 9039863) $ spack concretize --force -U
==> Error: Couldn't find patch for package builtin.hdf5 with sha256: 0b5dd6f233bc23532ccc79034a136605ec8c6a695f9d82fbcc3caf0c1c8c3119
```

Adam's above hack appears to resolve the issue.
